cloud-init: Support injection of additional SSH keys on instances (#14959)
canonical/lxd: doc/cloud-init.md

      - type: nameserver
        address: 10.10.10.254
```

## How to inject SSH keys on instances

> It is possible to inject SSH keys into LXD instances using `cloud-init.vendor-data` or `cloud-init.user-data`. For more information on how to do this using `cloud-config`, see [the cloud-init docs for SSH configuration](https://cloudinit.readthedocs.io/en/latest/reference/yaml_examples/ssh.html) for more information.

Reviewer: lets re-phrase this section to say something like:

- Its possible to inject ssh keys into LXD instances using cloud-init.
- This can be done using LXD's ``cloud-init.ssh-keys.` configuration options 
- Or via manually constructed  `cloud-init.vendor-data` or `cloud-init.user-data` (see cloud-init docs)

> Here is an example of a profile configuration that defines a key to be injected on the instance. This enables the owner of the private key to SSH into the instance as a user named `user`:

Reviewer 2: ```suggestion
The example profile configuration below defines a key to be injected on an instance. The injected key enables the owner of the private key to SSH into the instance as a user named `user`:
```

```yaml
config:
  cloud-init.vendor-data: |
    users:
      - name: user
        ssh_authorized_keys: ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJFDWcYmMrCZdk9JI29bAiHKD90oEUr8tqK5VvoO8Vcj
```

> It is also possible to inject an additional key for the user `user` on an instance that uses this profile in such a way that the key defined on the profile is still applied. This can be done using the configuration key `cloud-init.ssh-keys.<keyName>`.

Reviewer: into an instance

Reviewer: lets re-order this to promote the new simpler way of injecting ssh keys rather than having the older manual way first.

```bash
lxc launch <image> <instance-name> -c cloud-init.ssh-keys.<keyName> <public-key>
```

> Additional keys defined via `cloud-init.ssh-keys.<keyName>` may not be applied if LXD is not able to parse neither the existing `cloud-config.vendor-data` and `cloud-config.user-data` for that instance. This could happen, for example, if the content of those keys is a badly formatted YAML.

Reviewer 2: ```suggestion
Additional keys defined via `cloud-init.ssh-keys.<keyName>` cannot be applied if LXD is unable to parse either the existing `cloud-config.vendor-data` or `cloud-config.user-data` for that instance. This could happen, for example, if one of those keys contains badly formatted YAML.
```

      - type: nameserver
        address: 10.10.10.254
```

## How to inject SSH keys into instances

It is possible to inject SSH keys into LXD instances for an arbitrary user. This can be done using the configuration key `cloud-init.ssh-keys.<keyName>`.

> This configuration should respect the format `<user>:<key>`, where `<user` is a Linux username and `<key>` can be either a pure SSH public key or an import ID for a key hosted elsewhere. For example, `root:gh:githubUser` and `myUser:ssh-keyAlg publicKeyHash` are valid settings.

Reviewer 2: ```suggestion
Use the format `<user>:<key>` for its value, where `<user>` is a Linux username and `<key>` can be either a pure SSH public key or an import ID for a key hosted elsewhere. For example, `root:gh:githubUser` and `myUser:ssh-keyAlg publicKeyHash` are valid values.
```

> `cloud-init.ssh-keys.<keyName>` cannot be applied if LXD is unable to parse both the existing `cloud-config.vendor-data` and `cloud-config.user-data` for that instance. This could happen, for example, if those keys contain badly formatted YAML.

Reviewer 2: ```suggestion
`cloud-init.ssh-keys.<keyName>` cannot be applied if LXD is unable to parse both the existing {config:option}`instance-cloud-init:cloud-init.vendor-data` and {config:option}`instance-cloud-init:cloud-init.user-data` keys for that instance. This could happen, for example, if those keys contain badly formatted YAML.
```

> You can define SSH keys via `cloud-init.vendor-data` or `cloud-init.user-data` directly. Keys defined using `cloud-init.ssh-keys.<keyName>` do not conflict with those defined with either `cloud-init.vendor-data` or `cloud-init.user-data` in any way. For more information on how to use `cloud-config` to define SSH keys, see [the cloud-init docs for SSH configuration](https://cloudinit.readthedocs.io/en/latest/reference/yaml_examples/ssh.html).

Reviewer: This section would benefit from being expanded to explain that the SSH keys defined are merged into the config provided here and presented to the guest.

> It is also possible to define SSH keys via `cloud-init.vendor-data` or `cloud-init.user-data` directly. Defining keys using `cloud-init.ssh-keys.<keyName>` does not conflict with neither `cloud-init.vendor-data` or `cloud-init.user-data` in any way. For more information on how to do this using `cloud-config`, see [the cloud-init docs for SSH configuration](https://cloudinit.readthedocs.io/en/latest/reference/yaml_examples/ssh.html).

Reviewer 2: ```suggestion
You can define SSH keys via `cloud-init.vendor-data` or `cloud-init.user-data` directly. Keys defined using `cloud-init.ssh-keys.<keyName>` do not conflict with those defined with either `cloud-init.vendor-data` or `cloud-init.user-data` in any way. For more information on how to use `cloud-config` to define SSH keys, see [the cloud-init docs for SSH configuration](https://cloudinit.readthedocs.io/en/latest/reference/yaml_examples/ssh.html).
```

> Since `cloud-init` only runs on instance start, updates to `cloud-init.*` keys on a running instance will only take effect after restart.

Reviewer 2: ```suggestion
Since `cloud-init` only runs on instance start, updates to `cloud-init.*` keys on a running instance only take effect after restart.
```
